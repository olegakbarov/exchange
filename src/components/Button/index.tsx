import styled from "styled-components/macro";

const Button = styled.button`
  width: 100%;
  background-color: ${p => p.theme.buttonBgColor};
  padding: 10px 30px;
  border-radius: 4px;
  border: none;
  font-size: 16px;
  color: ${p => p.theme.white};

  &:hover {
    cursor: pointer;
  }
`;

export default Button;
