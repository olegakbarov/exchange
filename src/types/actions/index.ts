export * from './account';
// export * from './rates';