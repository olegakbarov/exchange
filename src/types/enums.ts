export enum ActionTypes {
  UpdateAccount = "UPDATE_ACCOUNT",
  UpdateRates = "UPDATE_RATES"
}

export type CurrencyCode =
  | "USD"
  | "EUR"
  | "GBP"
  | "AUD"
  | "BGN"
  | "BRL"
  | "CAD"
  | "CHF"
  | "CNY"
  | "CZK"
  | "DKK"
  | "HKD"
  | "HRK"
  | "HUF"
  | "IDR"
  | "ILS"
  | "INR"
  | "ISK"
  | "JPY"
  | "KRW"
  | "MXN"
  | "MYR"
  | "NOK"
  | "NZD"
  | "PHP"
  | "PLN"
  | "RON"
  | "RUB"
  | "SEK"
  | "SGD"
  | "THB"
  | "TRY"
  | "ZAR";
