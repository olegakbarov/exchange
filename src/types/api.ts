import { CurrencyCode } from "./enums";

export interface ApiResponse {
  base: CurrencyCode;
  rates: {
    [key in CurrencyCode]: number;
  };
}
