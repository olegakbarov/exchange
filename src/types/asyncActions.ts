import { AsyncActionTypes as AT } from './enums';

type ActionTemplate<T, P extends object> = {
  type: T;
  params: P;
};

export enum AsyncActionStatus {
  request = 'request',
  success = 'success',
  error = 'error'
}

export type AsyncRequest = () => Promise<Response>;

export type ARequest<T extends AT, P extends object = {}> = {
  status: AsyncActionStatus.request;
} & ActionTemplate<T, P>;

export type ASuccess<
  T extends AT,
  R extends object = {},
  P extends object = {}
> = {
  result: R;
  status: AsyncActionStatus.success;
} & ActionTemplate<T, P>;

export type AError<T extends AT, P extends object = {}> = {
  error: Error;
  status: AsyncActionStatus.error;
} & ActionTemplate<T, P>;

export type AsyncActions<
  T extends AT,
  R extends object = {},
  P extends object = {}
> = ARequest<T, P> | ASuccess<T, R, P> | AError<T, P>;

export type ToRequest<TWhere> = TWhere extends ARequest<infer T, infer P>
  ? ARequest<T, P>
  : never;

export type ToError<TWhere> = TWhere extends AError<infer T, infer P>
  ? AError<T, P>
  : never;

export type ToSuccess<TWhere> = TWhere extends ASuccess<
  infer T,
  infer R,
  infer P
>
  ? ASuccess<T, R, P>
  : never;
