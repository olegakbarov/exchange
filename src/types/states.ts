import { CurrencyCode } from "./enums";

export type AccountState = {
  [key in CurrencyCode]: number;
};

export type RatesState =
  | {
      [key in CurrencyCode]: {
        [key in CurrencyCode]: number;
      };
    }
  | {};

type State = AccountState;

export type RootState = {
  [key: string]: State;
};
