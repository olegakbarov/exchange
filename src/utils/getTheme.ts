import { Theme } from "../types/theme";
import { Themes } from "../types/enums";

export default function getTheme(name: Themes): Theme {
  switch (name) {
    case Themes.Light:
      return {
        white: "#fff",
        successGreen: "#56d256",
        errorRed: "#ff2828",
        bgColor: "#f8f8fa",
        fgColor: "#000",
        buttonBgColor: "rgb(38, 132, 255)",
        borderColor: "#eaeaea",
        inputBgColor: "#fff;"
      };

    case Themes.Dark:
      return {
        white: "#fff",
        successGreen: "#56d256",
        errorRed: "#ff2828",
        bgColor: "#000",
        fgColor: "#fff",
        buttonBgColor: "charcoal",
        borderColor: "#fff",
        inputBgColor: "#4d4d4d;"
      };
  }
}
