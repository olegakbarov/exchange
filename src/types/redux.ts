import { Action, AnyAction, Dispatch } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { RootState } from './states';

export type DA<T = void, A extends Action = AnyAction> = (
  dispatch: Dispatch<A>,
  getState: () => RootState
) => Promise<T>;

export type TDA<A extends Action = AnyAction> = ThunkDispatch<
  RootState,
  void,
  A
>;

export type Reducer<S, T extends Action> = (state: S, action: T) => S;

export type ReactDispatch<T> = T extends (...args: any[]) => Action
  ? T
  : T extends (
      ...args: infer A
    ) => (dispatch: any, getState: () => RootState) => infer X
  ? (...args: A) => X
  : never;
