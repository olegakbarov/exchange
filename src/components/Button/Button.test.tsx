import * as React from "react";
import { shallow, render } from "enzyme";
import "jest-styled-components";
import Button from "./index";
import getTheme from "../../utils/getTheme";
import { Themes } from "../../types/enums";

const BUTTON_TEXT = "Click";

describe("Button", () => {
  it("Shallow render snapshots are equal", () => {
    const tree = shallow(
      <Button theme={getTheme(Themes.Light)} onClick={() => {}}>
        {BUTTON_TEXT}
      </Button>
    );
    expect(tree).toMatchSnapshot();
  });

  it("Render snapshots are equal", () => {
    const tree = render(
      <Button theme={getTheme(Themes.Light)} onClick={() => {}}>
        {BUTTON_TEXT}
      </Button>
    );
    expect(tree).toMatchSnapshot();
  });
});
