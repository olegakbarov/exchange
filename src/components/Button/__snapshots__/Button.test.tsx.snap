// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render snapshots are equal 1`] = `
.c0 {
  width: 100%;
  background-color: rgb(38,132,255);
  padding: 10px 30px;
  border-radius: 4px;
  border: none;
  font-size: 16px;
  color: #fff;
}

.c0:hover {
  cursor: pointer;
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<button
  class="c0"
>
  Click
</button>
`;

exports[`Button Shallow render snapshots are equal 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Button-sc-14a3f2p-0",
        "isStatic": false,
        "rules": Array [
          "width:100%;background-color:",
          [Function],
          ";padding:10px 30px;border-radius:4px;border:none;font-size:16px;color:",
          [Function],
          ";&:hover{cursor:pointer;}&:disabled{opacity:0.5;cursor:not-allowed;}",
        ],
      },
      "displayName": "Button",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "Button-sc-14a3f2p-0",
      "target": "button",
      "toString": [Function],
      "usesTheme": true,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
  onClick={[Function]}
  theme={
    Object {
      "baseFontSize": "12px",
      "bgColor": "#f8f8fa",
      "borderColor": "#eaeaea",
      "buttonBgColor": "rgb(38, 132, 255)",
      "errorRed": "#ff2828",
      "fgColor": "#000",
      "inputBgColor": "#fff;",
      "shadowColor": "rgba(0, 0, 0, 0.3)",
      "successGreen": "#56d256",
      "white": "#fff",
    }
  }
>
  Click
</StyledComponent>
`;
